Gradle 1.7.0.CR1 regression: Gradle project does not work with test-fixtures plugin anymore

**Describe the bug**
In #10327 I was directed to use java-test-fixtures plugin for reused test resources. I am happily using it now. Unfortunately with 1.7.0.CR1 this stopped working. Reproducer can be found here: https://github.com/asodja/quarkus-test-fixtures-reproducer.

When running tests inside IDE (with Intellij Runner) I get:
```
java.lang.RuntimeException: java.io.UncheckedIOException: Error while reading file as JAR: /Users/anzes/prototype-dev/quarkus-test-fixtures-reproducible/build/libs/quarkus-test-fixtures-reproducible-my-version.jar
	at io.quarkus.test.junit.QuarkusTestExtension.throwBootFailureException(QuarkusTestExtension.java:508)
	at io.quarkus.test.junit.QuarkusTestExtension.beforeEach(QuarkusTestExtension.java:358)
...
Caused by: java.io.FileNotFoundException: /Users/anzes/prototype-dev/quarkus-test-fixtures-reproducible/build/libs/quarkus-test-fixtures-reproducible-my-version.jar (No such file or directory)
	at java.util.zip.ZipFile.open(Native Method)
	at java.util.zip.ZipFile.<init>(ZipFile.java:225)
...
```

And when I run `gradle test` from terminal I get:
```
MyResourceTest > testTestFixtures() FAILED
    java.lang.LinkageError: loader constraint violation: loader (instance of io/quarkus/bootstrap/classloading/QuarkusClassLoader) previously initiated loading for a different type with name "org/my/group/MyService"
        at java.lang.ClassLoader.defineClass1(Native Method)
        at java.lang.ClassLoader.defineClass(ClassLoader.java:757)
        at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:420)
        at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:363)
        at org.my.group.MyResourceTest.testTestFixtures(MyResourceTest.java:18)
```

Of course if I do `gradle build` and then run inside IntelliJ I get the same error as with `gradle test`, since jar is present in build directory.

**Expected behavior**
Test works and passes, no exception is thrown.

**Actual behavior**
Exception is thrown, test fails.


**To Reproduce**
Steps to reproduce the behavior:
1. git clone git@github.com:asodja/quarkus-test-fixtures-reproducer.git
2. Open in IntellIj and set Gradle: Build and run using: IntelliJ IDEA and Run tests using: IntelliJ IDEA
3. Run MyResourceTest, it fails
4. open terminal and run `gradle test`, it also fails
5. Set version of Quarkus to 1.6.1.Final
6. run `gradle test` again, it passes

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: MacBook Pro, Darwin Kernel Version 19.5.0
 - Output of `java -version`: openjdk version "1.8.0_242"
 - Quarkus version: 1.7.0.CR1
 - Build tool: Gradle 6.5.1 
---
Off topic, thank you for implementing #8737. As I follow there is still some work to do, like model caching etc, to have speed of running maven project from Intellij. But it looks like the first big step is done. Using IntelliJ runner really makes a difference when dealing with big projects, since Gradle initialization can be slow.

--- a/devtools/gradle/src/main/java/io/quarkus/gradle/builder/QuarkusModelBuilder.java
+++ b/devtools/gradle/src/main/java/io/quarkus/gradle/builder/QuarkusModelBuilder.java
@@ -28,7 +28,9 @@ import org.gradle.api.internal.artifacts.dependencies.DefaultExternalModuleDepen
 import org.gradle.api.plugins.Convention;
 import org.gradle.api.plugins.JavaPlugin;
 import org.gradle.api.plugins.JavaPluginConvention;
+import org.gradle.api.plugins.JavaTestFixturesPlugin;
 import org.gradle.api.tasks.SourceSet;
+import org.gradle.internal.component.external.model.TestFixturesSupport;
 import org.gradle.tooling.provider.model.ParameterizedToolingModelBuilder;
 
 import io.quarkus.bootstrap.BootstrapConstants;
@@ -100,12 +102,12 @@ public class QuarkusModelBuilder implements ParameterizedToolingModelBuilder {
         ArtifactCoords appArtifactCoords = new ArtifactCoordsImpl(project.getGroup().toString(), project.getName(),
                 project.getVersion().toString());
 
-        return new QuarkusModelImpl(new WorkspaceImpl(appArtifactCoords, getWorkspace(project.getRootProject())),
+        return new QuarkusModelImpl(new WorkspaceImpl(appArtifactCoords, getWorkspace(project.getRootProject(), mode)),
                 new HashSet<>(appDependencies.values()),
                 extensionDependencies);
     }
 
-    public Set<WorkspaceModule> getWorkspace(Project project) {
+    public Set<WorkspaceModule> getWorkspace(Project project, LaunchMode mode) {
         Set<WorkspaceModule> modules = new HashSet<>();
         for (Project subproject : project.getAllprojects()) {
             final Convention convention = subproject.getConvention();
@@ -113,18 +115,26 @@ public class QuarkusModelBuilder implements ParameterizedToolingModelBuilder {
             if (javaConvention == null) {
                 continue;
             }
-            modules.add(getWorkspaceModule(subproject));
+            modules.add(getWorkspaceModule(subproject, mode));
         }
         return modules;
     }
 
-    private WorkspaceModule getWorkspaceModule(Project project) {
+    private WorkspaceModule getWorkspaceModule(Project project, LaunchMode mode) {
         ArtifactCoords appArtifactCoords = new ArtifactCoordsImpl(project.getGroup().toString(), project.getName(),
                 project.getVersion().toString());
         final SourceSet mainSourceSet = QuarkusGradleUtils.getSourceSet(project, SourceSet.MAIN_SOURCE_SET_NAME);
+        final SourceSetImpl modelSourceSet = convert(mainSourceSet);
+
+        // If the project exposes test fixtures, we must add them as source directory
+        if (mode.equals(LaunchMode.TEST) && project.getPlugins().hasPlugin(JavaTestFixturesPlugin.class)) {
+            final SourceSet fixtureSourceSet = QuarkusGradleUtils.getSourceSet(project,
+                    TestFixturesSupport.TEST_FIXTURE_SOURCESET_NAME);
+            modelSourceSet.addSourceDirectories(fixtureSourceSet.getOutput().getClassesDirs().getFiles());
+        }
 
         return new WorkspaceModuleImpl(appArtifactCoords, project.getProjectDir().getAbsoluteFile(),
-                project.getBuildDir().getAbsoluteFile(), getSourceSourceSet(mainSourceSet), convert(mainSourceSet));
+                project.getBuildDir().getAbsoluteFile(), getSourceSourceSet(mainSourceSet), modelSourceSet);
     }
 
     private Set<org.gradle.api.artifacts.Dependency> getEnforcedPlatforms(Project project) {
@@ -239,7 +249,6 @@ public class QuarkusModelBuilder implements ParameterizedToolingModelBuilder {
                     a.getId().getComponentIdentifier() instanceof ProjectComponentIdentifier) {
                 Project projectDep = project.getRootProject()
                         .findProject(((ProjectComponentIdentifier) a.getId().getComponentIdentifier()).getProjectPath());
-
                 dep = toDependency(a, configurationName, projectDep);
             } else {
                 dep = toDependency(a, configurationName);

--- a/independent-projects/bootstrap/gradle-resolver/src/main/java/io/quarkus/bootstrap/resolver/model/impl/SourceSetImpl.java
+++ b/independent-projects/bootstrap/gradle-resolver/src/main/java/io/quarkus/bootstrap/resolver/model/impl/SourceSetImpl.java
@@ -3,17 +3,22 @@ package io.quarkus.bootstrap.resolver.model.impl;
 import io.quarkus.bootstrap.resolver.model.SourceSet;
 import java.io.File;
 import java.io.Serializable;
+import java.util.HashSet;
 import java.util.Set;
 import java.util.stream.Collectors;
 
 public class SourceSetImpl implements SourceSet, Serializable {
 
-    private final Set<File> sourceDirectories;
+    private Set<File> sourceDirectories = new HashSet<>();
     private final File resourceDirectory;
 
     public SourceSetImpl(Set<File> sourceDirectories, File resourceDirectory) {
-        this.sourceDirectories = sourceDirectories;
+        this.sourceDirectories.addAll(sourceDirectories);
         this.resourceDirectory = resourceDirectory;
+    }
+
+    public void addSourceDirectories(Set<File> files) {
+        sourceDirectories.addAll(files);
     }
 
     @Override

--- a/integration-tests/gradle/src/test/java/io/quarkus/gradle/TestFixtureModuleTest.java
+++ b/integration-tests/gradle/src/test/java/io/quarkus/gradle/TestFixtureModuleTest.java
@@ -1,0 +1,21 @@
+package io.quarkus.gradle;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URISyntaxException;
+
+import org.junit.jupiter.api.Test;
+
+public class TestFixtureModuleTest extends QuarkusGradleWrapperTestBase {
+
+    @Test
+    public void testTaskShouldUseTestFixtures() throws IOException, URISyntaxException, InterruptedException {
+        final File projectDir = getProjectDir("test-fixtures-module");
+
+        final BuildResult result = runGradleWrapper(projectDir, "clean", "test");
+
+        assertThat(result.getTasks().get(":test")).isEqualTo(BuildResult.SUCCESS_OUTCOME);
+    }
+}

--- a/integration-tests/gradle/src/test/resources/test-fixtures-module/build.gradle
+++ b/integration-tests/gradle/src/test/resources/test-fixtures-module/build.gradle
@@ -1,0 +1,35 @@
+plugins {
+    id 'java'
+    id 'java-test-fixtures'
+    id 'io.quarkus'
+}
+
+repositories {
+     mavenLocal()
+     mavenCentral()
+}
+
+dependencies {
+    implementation 'io.quarkus:quarkus-resteasy-jsonb'
+    implementation enforcedPlatform("${quarkusPlatformGroupId}:${quarkusPlatformArtifactId}:${quarkusPlatformVersion}")
+    implementation 'io.quarkus:quarkus-resteasy'
+
+    testImplementation 'io.quarkus:quarkus-junit5'
+    testImplementation 'io.rest-assured:rest-assured'
+}
+
+group 'my-groupId'
+version 'my-version'
+
+compileJava {
+    options.encoding = 'UTF-8'
+    options.compilerArgs << '-parameters'
+}
+
+compileTestJava {
+    options.encoding = 'UTF-8'
+}
+
+test {
+    systemProperty "java.util.logging.manager", "org.jboss.logmanager.LogManager"
+}

--- a/integration-tests/gradle/src/test/resources/test-fixtures-module/gradle.properties
+++ b/integration-tests/gradle/src/test/resources/test-fixtures-module/gradle.properties
@@ -1,0 +1,4 @@
+#Quarkus Gradle TS
+#Sat May 23 23:11:14 CEST 2020
+quarkusPlatformArtifactId=quarkus-bom
+quarkusPlatformGroupId=io.quarkus

--- a/integration-tests/gradle/src/test/resources/test-fixtures-module/settings.gradle
+++ b/integration-tests/gradle/src/test/resources/test-fixtures-module/settings.gradle
@@ -1,0 +1,15 @@
+pluginManagement {
+    repositories {
+        jcenter()
+        mavenLocal()
+        mavenCentral()
+        gradlePluginPortal()
+    }
+    //noinspection GroovyAssignabilityCheck
+    plugins {
+        id 'io.quarkus' version "${quarkusPluginVersion}"
+    }
+}
+
+rootProject.name='io.quarkus.reproducer'
+

--- a/integration-tests/gradle/src/test/resources/test-fixtures-module/src/main/java/org.my.group/MyFactory.java
+++ b/integration-tests/gradle/src/test/resources/test-fixtures-module/src/main/java/org.my.group/MyFactory.java
@@ -1,0 +1,7 @@
+package org.my.group;
+
+public interface MyFactory {
+
+    MyService createMyService();
+
+}

--- a/integration-tests/gradle/src/test/resources/test-fixtures-module/src/main/java/org.my.group/MyResource.java
+++ b/integration-tests/gradle/src/test/resources/test-fixtures-module/src/main/java/org.my.group/MyResource.java
@@ -1,0 +1,16 @@
+package org.my.group;
+
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+
+@Path("/hello")
+public class MyResource {
+
+    @GET
+    @Produces(MediaType.TEXT_PLAIN)
+    public String hello() {
+        return "hello";
+    }
+}

--- a/integration-tests/gradle/src/test/resources/test-fixtures-module/src/main/java/org.my.group/MyService.java
+++ b/integration-tests/gradle/src/test/resources/test-fixtures-module/src/main/java/org.my.group/MyService.java
@@ -1,0 +1,7 @@
+package org.my.group;
+
+public interface MyService {
+
+    String getSomething();
+
+}

--- a/integration-tests/gradle/src/test/resources/test-fixtures-module/src/test/java/org.my.group/MyResourceTest.java
+++ b/integration-tests/gradle/src/test/resources/test-fixtures-module/src/test/java/org.my.group/MyResourceTest.java
@@ -1,0 +1,20 @@
+package org.my.group;
+
+import io.quarkus.test.junit.QuarkusTest;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+import com.example.MyTestFactory;
+
+@QuarkusTest
+public class MyResourceTest {
+
+    @Test
+    public void testTestFixtures() {
+        MyService myService = new MyTestFactory().createMyService();
+
+        String result = myService.getSomething();
+
+        Assertions.assertEquals("test-fixtures", result);
+    }
+}

--- a/integration-tests/gradle/src/test/resources/test-fixtures-module/src/testFixtures/java/org.my.group/MyTestFactory.java
+++ b/integration-tests/gradle/src/test/resources/test-fixtures-module/src/testFixtures/java/org.my.group/MyTestFactory.java
@@ -1,0 +1,13 @@
+package com.example;
+
+import org.my.group.MyFactory;
+import org.my.group.MyService;
+
+public class MyTestFactory implements MyFactory {
+
+    @Override
+    public MyService createMyService() {
+        return () -> "test-fixtures";
+    }
+
+}

--- a/test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java
+++ b/test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java
@@ -8,6 +8,7 @@ import java.io.IOException;
 import java.lang.reflect.Constructor;
 import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
+import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.Paths;
 import java.util.AbstractMap;
@@ -130,6 +131,10 @@ public class QuarkusTestExtension
                     rootBuilder.add(testResourcesLocation);
                 }
             }
+            if (Files.exists(testClassLocation.getParent().resolve("testFixtures"))) {
+                rootBuilder.add(testClassLocation.getParent().resolve("testFixtures"));
+            }
+
             originalCl = Thread.currentThread().getContextClassLoader();
             Map<String, String> sysPropRestore = new HashMap<>();
             sysPropRestore.put(ProfileManager.QUARKUS_TEST_PROFILE_PROP,

